metric_learn/scml.py

```diff
@@ -0,0 +1,596 @@
+
+    w = np.zeros((1, n_basis))
+    avg_grad_w = np.zeros((1, n_basis))
+
+    best_obj = np.inf
+
+    rng = check_random_state(self.random_state)
+    rand_int = rng.randint(low=0, high=sizeT, size=self.max_iter)
+    for iter in range(self.max_iter):
+      if (iter % self.output_iter == 0):
+        # regularization part of obj function
+        obj1 = np.sum(w)*self.beta
+
+      # Every triplet distance difference in the space given by L
+      # plus a slack of one
+        slack_val = 1 + np.matmul(dist_diff, w.T)
+      # Mask of places with positive slack
+        slack_mask = slack_val > 0
+
+        # loss function of learning task part of obj function
+        obj2 = sum_where(slack_val, slack_mask)/sizeT
+
+        obj = obj1 + obj2
+        if(self.verbose):
+          count = np.sum(slack_mask)
+          print("[Global] iter %d\t obj %.6f\t num_imp %d" % (iter,
+                obj, count))
+
+        # update the best
+        if (obj < best_obj):
+          best_obj = obj
+          best_w = w
+
+      # TODO:
+      # Maybe allow the usage of mini-batch opt?
+
+      idx = rand_int[iter]
+
+      slack_val = 1 + np.matmul(dist_diff[idx, :], w.T)
+
+      if (slack_val > 0):
+        avg_grad_w = (iter * avg_grad_w + dist_diff[idx, :]) / (iter+1)
+      else:
+        avg_grad_w = iter * avg_grad_w / (iter+1)
+
+      scale_f = -np.sqrt(iter+1) / self.gamma
+
+      # proximal operator with negative trimming equivalent
+      w = scale_f * np.minimum(avg_grad_w + self.beta, 0)
+
+    if(self.verbose):
+      print("max iteration reached.")
+
+    # return L matrix yielded from best weights
+    self.components_ = self._get_components(best_w, basis)
+
+    return self
+
+  def _compute_dist_diff(self, T, X, basis):
+    """
+    Helper function to compute the distance difference of every triplet in the
+    space yielded by the basis set.
+    """
+    # Transformation of data by the basis set
+    XB = np.matmul(X, basis.T)
+
+    lenT = len(T)
+    # get all positive and negative pairs with lowest index first
+    # np.array (2*lenT,2)
+    T_pairs_sorted = np.sort(np.vstack((T[:, [0, 1]], T[:, [0, 2]])),
+                             kind='stable')
+    # calculate all unique pairs and their indices
+    uniqPairs, indices = np.unique(T_pairs_sorted, return_inverse=True,
+                                   axis=0)
+    # calculate L2 distance acording to bases only for unique pairs
+    dist = np.square(XB[uniqPairs[:, 0], :]-XB[uniqPairs[:, 1], :])
+
+    # return the diference of distances between all positive and negative
+    # pairs
+    return dist[indices[:lenT]]-dist[indices[lenT:]]
+
+  def _get_components(self, w, basis):
+    """
+    get components matrix (L) from computed mahalanobis matrix
+    """
+
+    # get rid of inactive bases
+    active_idx, = w > 0
+    w = w[..., active_idx]
+    basis = basis[active_idx, :]
+
+    K, d = basis.shape
+
+    if(K < d):  # if metric is low-rank
```

Comment:
in this case we are currently returning a rectangular `L` of dimension `(K, d)`, correct? If so we should consider printing something to let the user know that the learned transformation reduces dimension

Reply:
Yes, you are correct! Should this be a warning? Also, do you think its a good idea to change `K, d` to `n_basis, n_features`?

Reply:
Definitely. Yes, a warning sounds good I guess?